design/MultipleQueues.md

```diff
@@ -0,0 +1,24 @@
+# Multiple Queues
+
+Queues are created at Device creation time, and are exposed via `sequence<Queue> Device.queues`.
+There are multiple types of queues, denoted by the `.type` bitfield on the Queue object.
+Queues can support `graphics` operations, `compute` operations, both, or neither.
+"Neither" Queues can only execute copy commands.
+If there are any queues that support graphics, there is at least one queue that supports
+both compute and graphics.
+The most capable queue is always first in the `Device.queues` list.
+(It will either be a graphics+compute queue, or compute queue)
+Devices must be able to support at least a compute queue, or device creation will fail.
+
+## Synchronization
+
+Queues can have fences inserted into them, and any queue can wait on that fence to be complete.
+Synchronizing usage of resources across queues should be done by telling a queue to wait until
+a fence is complete, followed by submitting command buffers that depend on that fence.
+
+## Further work
+
+- Transfering resources between queues, particularly for VK_SHARING_MODE_EXCLUSIVE, as opposed to _CONCURRENT.
+- App-requested number of same-type queues, instead of implementation-dictated-only.
```

Comment:
I don't think queues map to GPU cores, any GPU core can execute any queue's work... at least in Vulkan.

Reply:
@devshgraphicsprogramming right. Can you come up with a different example on when exposing the available queue types/counts would make the user code run better (which is the point @RafaelCintron is making)?

Reply:
I think the nicely illustrated example in this post https://mynameismjp.wordpress.com/2018/06/17/breaking-down-barriers-part-3-multiple-command-processors/ as well as the preceeding post, illustrate the need for multiple queues from the same family

Reply:
The question here is not whether or not to have multiple queues (PR is all about multiple queues). The question is whether we should expose what is available, or just let the user tell us what they need and want to work with.

Reply:
I address that better in my other reply https://github.com/gpuweb/gpuweb/pull/95#discussion_r234278773

I don't think you want to put yourself in the position of the scheduler and dependency tracker, you're pretty much making and doing what an enhanced OpenGL implementation would do.

Reply:
Also the example provided by @RafaelCintron is a bit moot
> Suppose you have a fictional system with 6 compute cores. If the developer submits six dispatch calls with no dependencies between them, the driver is free to run each operation on a different core, utilizing all six of them. Similarly, the driver can also utilize the same 6 cores if the developer submits a single dispatch call with maxThreads=6. In these two instances, queuing things on compute queues 2-6 will do the developer no good.

A single queue will fill up all of your Compute Units / Shader Processors with work anyway, so that `maxThreads` parameter is moot anyway.
In the absence of dependencies as in @RafaelCintron's example a single queue and command processor will keep your entire GPU with all of its CU/SP/SMs busy with no idle time (bar a context switch between compute and graphics on old NVIDIAs).

Reply:
You only need as many queues as the maximal slice in your unweighted Work-DAG, I forgot my high-school discrete math... but basically looks like maximum-cut which is NP-complete.